File: app/src/test/java/com/osfit/app/domain/PesosPropiosTest.kt
package com.osfit.app.domain

import com.osfit.app.data.model.Ejercicio
import org.junit.Assert.assertEquals
import org.junit.Test

/** Unidad: los pesos que el entrenador le pone a una clienta encima de la plantilla. */
class PesosPropiosTest {

    private fun ejercicio(nombre: String, peso: String = "") =
        Ejercicio(nombre = nombre, series = 4, repeticiones = "10", pesoONota = peso)

    @Test
    fun `sin pesos propios devuelve los de la plantilla`() {
        val ejercicios = listOf(ejercicio("Press banca", "30 kg"))
        assertEquals(ejercicios, conPesosPropios(ejercicios, emptyMap()))
    }

    @Test
    fun `el peso propio pisa al de la plantilla`() {
        val ejercicios = listOf(ejercicio("Press banca", "30 kg"))

        val resultado = conPesosPropios(ejercicios, mapOf("press banca" to "40 kg"))

        assertEquals("40 kg", resultado.single().pesoONota)
    }

    @Test
    fun `solo cambia el ejercicio que tiene peso propio`() {
        val ejercicios = listOf(ejercicio("Press banca", "30 kg"), ejercicio("Remo", "25 kg"))

        val resultado = conPesosPropios(ejercicios, mapOf("press banca" to "40 kg"))

        assertEquals("40 kg", resultado[0].pesoONota)
        assertEquals("el otro se queda con el de la plantilla", "25 kg", resultado[1].pesoONota)
    }

    @Test
    fun `la busqueda ignora mayusculas y espacios de sobra`() {
        val ejercicios = listOf(ejercicio("  Press   Banca "))

        val resultado = conPesosPropios(ejercicios, mapOf("press banca" to "40 kg"))

        assertEquals("40 kg", resultado.single().pesoONota)
    }

    @Test
    fun `un peso propio en blanco cae al de la plantilla`() {
        val ejercicios = listOf(ejercicio("Press banca", "30 kg"))

        val resultado = conPesosPropios(ejercicios, mapOf("press banca" to "   "))

        assertEquals("30 kg", resultado.single().pesoONota)
    }

    @Test
    fun `reordenar la plantilla no despega los pesos`() {
        // Es la razón de indexar por nombre y no por posición.
        val original = listOf(ejercicio("Press banca"), ejercicio("Remo"))
        val reordenada = listOf(ejercicio("Remo"), ejercicio("Press banca"))
        val propios = mapOf("press banca" to "40 kg")

        assertEquals("40 kg", conPesosPropios(original, propios)[0].pesoONota)
        assertEquals("40 kg", conPesosPropios(reordenada, propios)[1].pesoONota)
    }

    @Test
    fun `guardar normaliza la clave`() {
        assertEquals(
            mapOf("press banca" to "40 kg"),
            conPesoPropio(emptyMap(), "  Press   BANCA ", "40 kg")
        )
    }

    @Test
    fun `guardar en blanco quita la entrada en vez de dejarla vacia`() {
        val propios = mapOf("press banca" to "40 kg")

        assertEquals(emptyMap<String, String>(), conPesoPropio(propios, "Press banca", ""))
    }

    @Test
    fun `un nombre vacio no ensucia el mapa`() {
        val propios = mapOf("press banca" to "40 kg")

        assertEquals(propios, conPesoPropio(propios, "   ", "40 kg"))
    }
}

File: web/src/pesosPropios.test.ts
import { describe, expect, it } from "vitest";
import type { Ejercicio } from "./datos";
import { claveEjercicio, conPesosPropios } from "./pesosPropios";

function ejercicio(nombre: string, pesoONota = ""): Ejercicio {
  return { nombre, series: 4, repeticiones: "10", pesoONota };
}

describe("conPesosPropios", () => {
  it("sin pesos propios devuelve los de la plantilla", () => {
    const ejercicios = [ejercicio("Press banca", "30 kg")];
    expect(conPesosPropios(ejercicios, {})).toEqual(ejercicios);
  });

  it("el peso propio pisa al de la plantilla", () => {
    const resultado = conPesosPropios([ejercicio("Press banca", "30 kg")], {
      "press banca": "40 kg",
    });
    expect(resultado[0].pesoONota).toBe("40 kg");
  });

  it("solo cambia el ejercicio que tiene peso propio", () => {
    const resultado = conPesosPropios(
      [ejercicio("Press banca", "30 kg"), ejercicio("Remo", "25 kg")],
      { "press banca": "40 kg" }
    );
    expect(resultado[0].pesoONota).toBe("40 kg");
    expect(resultado[1].pesoONota).toBe("25 kg");
  });

  it("la búsqueda ignora mayúsculas y espacios de sobra", () => {
    const resultado = conPesosPropios([ejercicio("  Press   Banca ")], {
      "press banca": "40 kg",
    });
    expect(resultado[0].pesoONota).toBe("40 kg");
  });

  it("un peso propio en blanco cae al de la plantilla", () => {
    const resultado = conPesosPropios([ejercicio("Press banca", "30 kg")], {
      "press banca": "   ",
    });
    expect(resultado[0].pesoONota).toBe("30 kg");
  });

  it("un mapa ausente no rompe", () => {
    // Firestore omite los campos que nunca se escribieron.
    const ejercicios = [ejercicio("Press banca", "30 kg")];
    expect(conPesosPropios(ejercicios, undefined)).toEqual(ejercicios);
  });

  it("reordenar la plantilla no despega los pesos", () => {
    const propios = { "press banca": "40 kg" };
    const reordenada = [ejercicio("Remo"), ejercicio("Press banca")];
    expect(conPesosPropios(reordenada, propios)[1].pesoONota).toBe("40 kg");
  });

  it("la clave normaliza igual que el gemelo de Kotlin", () => {
    expect(claveEjercicio("  Press   BANCA ")).toBe("press banca");
  });
});
